src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -1100,10 +1106,114 @@ int32_t SystemNative_ReadLink(const char* path, char* buffer, int32_t bufferSize
     return (int32_t)count;
 }
 
-int32_t SystemNative_Rename(const char* oldPath, const char* newPath)
+// Potential issue: renaming a file onto another link to itself can decrease the total link count of the file.
+// We could only fix this if there were a way to ask the filesystem if two dnodes were identical, but there isn't.
+static int32_t RenameOnEexist(int32_t oldResult, struct stat *targetBuf, const char *oldPath, const char *newPath)
+{
+    struct stat oldBuf;
+    struct stat newBuf;
+    bool haveTarget = targetBuf != NULL;
+    if (!targetBuf)
+    {
+        targetBuf = &newBuf;
+    }
+    int savedErrno = errno;
+    errno = 0;
+    if (!lstat(oldPath, &oldBuf) && (haveTarget || !lstat(newPath, targetBuf))
+            && oldBuf.st_dev == targetBuf->st_dev && oldBuf.st_ino == targetBuf->st_ino)
+    {
+        errno = 0;
+        return rename(oldPath, newPath);
+    }
+    if (errno == EINTR)
+    {
+        return -1;
+    }
+    errno = savedErrno;
+    return oldResult;
+}
+
+int32_t SystemNative_Rename(const char* oldPath, const char* newPath, int32_t flags)
 {
     int32_t result;
-    while ((result = rename(oldPath, newPath)) < 0 && errno == EINTR);
+    if (flags & 1)
+    {
+        while ((result = rename(oldPath, newPath)) < 0 && errno == EINTR);
+    }
+    else
+    {
+#if HAVE_RENAMEAT2 && defined(RENAME_NOREPLACE)
+        // Linux renameat2() provides the functionality we want inline.
+        // This is in a feature check block as this API may well be ported to other OSes later.
+        do {
+            result = renameat2(AT_FDCWD, oldPath, AT_FDCWD, newPath, RENAME_NOREPLACE);
```

Comment:
I'm calling renameat2() for RENAME_NOREPLACE. AT_FDCWD is what you pass if you don't have anything else to pass in that slot.

Reply:
On third attempt, pronoun [it] is ambiguous. Do you mean AT_FDCWD or RENAME_NOREPLACE? A unit test for missing RENAME_NOREPLACE would just check that MoveFile(,,false) doesn't ovewrite its target and I think that already exists. If AT_FDCWD isn't passed, it will fail pretty hard as 0 won't be a directory.